src/xopen/__init__.py

```diff
@@ -243,11 +246,45 @@ def __next__(self):
         raise io.UnsupportedOperation('not readable')
 
 
+class DummyReader(Thread):
```

Comment:
I am not a fan of using threads. These have a tendency to create problems faster than that they solve them. If `terminate()` blocks, send_signal can be used instead: https://docs.python.org/3.7/library/subprocess.html#subprocess.Popen.send_signal . And then use communicate or wait. Both have timeouts.

Reply:
One problem with `communicate()` is that it will close `self._stderr`, so if we want the error message we need to do `self._stderr.read()` _before_  `communicate()`. However, `read()` will block untill the EOF, meaning we need to process to be terminated for which we need `wait()` or `communicate()` first. Any suggestions?

Reply:
`terminate()` just calls `send_signal(SIGTERM)` on Unix https://github.com/python/cpython/blob/115dea9e2602b96b63390f00cc880e90c433efa2/Lib/subprocess.py#L2092 so that should be enough.

@DriesSchaumont Can you say why you added the `DummyReader` class at all? The tests passed even before you added it. And the more I look at it, the less I understand why `wait()` would actually block. If `pbzip2` gets a SIGTERM while blocked on the output pipe, it still terminates correctly.

Reply:
`communicate()` returns what it reads from stdin and stderr (as a tuple), so you could pass it on to to `_raise_if_error`.

But as I said above, I don’t get why it’s necessary at all.

Reply:
@marcelm The problem is that pbzip2 is not killed by SIGTERM, which was causing the hanging tests (`test_reader_close`). The tests passed because I first implemented a solution where SIGKILL would be issued if SIGTERM failed to kill the process. I did not like this particular solution because if we use SIGKILL there is not way to distinguish between for example the kernel killing the process or xopen using  `kill()`. It also used a timeout, but different programs have different ways to handle SIGTERM so choosing a timer value seemed arbitrary...   I then noticed that pbzip2 does react to SIGTERM when the PIPE buffers are no longer full, which is why I implemented the seperate thread to flush it. 

As to _why_ pbzip2 is not closed by SIGTERM in this case, I am not sure... I will try to `communicate()` to see if I can avoid using custom thread.

Reply:
I removed the DummyReader in favor of using  `communicate()`. Closing this thread, please reopen if we should discuss this more.

Reply:
Ah, so are you on macOS? For me (on Linux), `pbzip2` does exit on SIGTERM.

But anyway, if `communicate()` works, then we have a solution.